### PR TITLE
deploy-mediawiki: add arguments for branch and rebuildLC language

### PR DESCRIPTION
**Pull request by Universal-Omega**

### CODE REVIEW

File: modules/mediawiki/files/bin/deploy-mediawiki.py [hunk 1]
```diff
@@ -149,11 +149,15 @@ def _construct_rsync_command(time: str, dest: str, recursive: bool = True, local
     raise Exception(f'Error constructing command. Either server was missing or {location} != {dest}')  # noqa: R503
 
 
-def _construct_git_pull(repo: str, submodules: bool = False) -> str:
+def _construct_git_pull(repo: str, submodules: bool = False, branch: Union[bool, str] = False) -> str:
     if submodules:
```
Comment (RhinosF1) [inline modules/mediawiki/files/bin/deploy-mediawiki.py:153]: Why? I assume you're trying to except False. Use None and type check it to be optional.

File: modules/mediawiki/files/bin/deploy-mediawiki.py [hunk 2]
```diff
@@ -188,13 +192,18 @@ def run(args: argparse.Namespace, start: float) -> None:
         if args.world and 'world' not in pull:
             pull.append('world')
         if pull:
+            if args.branch:
+                branch = args.branch
+            else:
+                branch = False
```
Comment (RhinosF1) [inline modules/mediawiki/files/bin/deploy-mediawiki.py:198]: Use None not False

File: modules/mediawiki/files/bin/deploy-mediawiki.py [hunk 3]
```diff
@@ -188,13 +192,18 @@ def run(args: argparse.Namespace, start: float) -> None:
...
+
             for repo in pull:
                 if repo == 'world':
                     sm = True
                 else:
                     sm = False
                 try:
-                    stage.append(_construct_git_pull(repo, submodules=sm))
+                    stage.append(_construct_git_pull(repo, submodules=sm, branch=branch))
```
Comment (RhinosF1) [inline modules/mediawiki/files/bin/deploy-mediawiki.py:206]: Why not use args.branch directly?

File: modules/mediawiki/files/bin/deploy-mediawiki.py [hunk 4]
```diff
@@ -226,8 +235,13 @@ def run(args: argparse.Namespace, start: float) -> None:
             exitcodes.append(run_command(cmd))
         non_zero_code(exitcodes)
         if args.l10n:  # setup l10n
+            if args.lang:
+                lang = f'--lang={args.lang}'
```
Comment (RhinosF1) [inline modules/mediawiki/files/bin/deploy-mediawiki.py:239]: What happens if it's an invalid language?

**Review by RhinosF1 — CHANGES_REQUESTED** [review]

**Review by RhinosF1 — CHANGES_REQUESTED** [review]
Please add tests for new functionality

**Comment by Universal-Omega** [comment]
> Please add tests for new functionality

Python tests are above my ability at the moment. While I could probably figure it out, I don't feel the need for me to more than I need to for this PR. This PR is supposed to make it easier to pull stuff, and faster to run world when only some stuff needs done (like just English for example). If I have to add python tests everytime I change something to it, I just won't change anything, and will abandon this PR. It is not worth my time. If you really want tests for this, then please submit a PR to my forked branch, or a follow-up PR to this. I'm not interested in learning how to make/update python tests at the moment with to much other things going on in RL for me, and other more pressing issues at Miraheze then that. This PR was supposed to be something simple but making it easier at times to use. I don't want it to be more complicated for me right now, and just don't have time to do that now.

I'm requesting that this review be retracted, and this PR still merged without tests, or that you submit a PR to my fork to add them. Thanks!

**Comment by RhinosF1** [comment]
> > Please add tests for new functionality
> 
> Python tests are above my ability at the moment. While I could probably figure it out, I don't feel the need for me to more than I need to for this PR. This PR is supposed to make it easier to pull stuff, and faster to run world when only some stuff needs done (like just English for example). If I have to add python tests everytime I change something to it, I just won't change anything, and will abandon this PR. It is not worth my time. If you really want tests for this, then please submit a PR to my forked branch, or a follow-up PR to this. I'm not interested in learning how to make/update python tests at the moment with to much other things going on in RL for me, and other more pressing issues at Miraheze then that. This PR was supposed to be something simple but making it easier at times to use. I don't want it to be more complicated for me right now, and just don't have time to do that now.

It's about 3 lines for the git changes which is what I care about and yes python tests should be added as much as we can. This is a critical part of our infra. I can happily write them for you but I'm going to block any PR that changes functionality on our team's python scripts without adding tests for the change.
> I'm requesting that this review be retracted, and this PR still merged without tests, or that you submit a PR to my fork to add them. Thanks!

**Comment by Reception123** [comment]
A few observations about the comments here:

1) Regarding the mandatory tests for every new functionality, I don't personally think this should be absolutely mandatory. We are quite a small organisation/team and sometimes we need to prioritise functionality over testing. If you believe that testing is very important, then there is nothing stopping you from adding it after / modifying the PR.

2) Regarding the statement "I'm going to block any PR that changes functionality " I think there's a few issues here. The statement seems to imply that a single member of the MW team can simply decide to block a change - I don't think a single member should be able to since we are a collaborative team. Either way, there is clearly a structure with the EM being the one who would be able to 'block' changes, especially in this case since I am the only one who can merge the change from the team. Also an issue is that there is really no policy or agreement about mandating tests, so without a clear rule I don't see why it should just be decided like that unilaterally.

Rather than asking people to do something, it's probably best to do it yourself if you want to see it done. My view however is that without a clear policy/rule, I don't see why this shouldn't be merged and why testing should be prioritized over functionality. If you want  such a rule/policy, that should be discussed with the team, as there is clearly disagreement about the such a rule.
Please do also consider your choice of words which seem to imply that you have more 'power' than other members of the team.

File: modules/mediawiki/files/bin/deploy-mediawiki.py [hunk 5]
```diff
@@ -149,11 +150,15 @@ def _construct_rsync_command(time: str, dest: str, recursive: bool = True, local
...
 
-def _construct_git_pull(repo: str, submodules: bool = False) -> str:
+def _construct_git_pull(repo: str, submodules: bool = False, branch: Optional[str] = None) -> str:
     if submodules:
         extrap = '--recurse-submodules'
     else:
         extrap = ''
+
+    if branch:
+        extrap += f'--branch={branch}'
```
Comment (RhinosF1) [inline modules/mediawiki/files/bin/deploy-mediawiki.py:160]: Tests added. Please see output as --recurse-submodules & --branch don't work. That's why we do tests and this method is supposed to be easy to check.

**Review by RhinosF1 — CHANGES_REQUESTED** [review]
See comments, doesn't work.

**Comment by RhinosF1** [comment]
https://github.com/Universal-Omega/puppet/pull/7/files

File: .github/pythonchecks.txt [hunk 6]
```diff
@@ -18,3 +18,4 @@ mypy==0.931
 types-requests==2.27.7
 pytest-xdist==2.5.0
 requests==2.27.1
```
Comment (RhinosF1) [inline .github/pythonchecks.txt:20]: This doesn't exist in Debian. We'll need to somehow install it on the appservers via puppet.

File: modules/mediawiki/files/bin/deploy-mediawiki.py [hunk 7]
```diff
@@ -226,8 +231,17 @@ def run(args: argparse.Namespace, start: float) -> None:
...
+                    if not tag_is_valid(language):
+                        raise ValueError(f'{language} is not a valid language.')
+
+                lang = f'--lang={args.lang}'
+            else:
+                lang = ''
+
             postinstall.append(f'sudo -u www-data php /srv/mediawiki/w/maintenance/mergeMessageFileList.php --quiet --wiki={envinfo["wikidbname"]} --output /srv/mediawiki/config/ExtensionMessageFiles.php')
-            rebuild.append(f'sudo -u www-data php /srv/mediawiki/w/maintenance/rebuildLocalisationCache.php --quiet --wiki={envinfo["wikidbname"]}')
+            rebuild.append(f'sudo -u www-data php /srv/mediawiki/w/maintenance/rebuildLocalisationCache.php {lang} --quiet --wiki={envinfo["wikidbname"]}')
```
Comment (RhinosF1) [inline modules/mediawiki/files/bin/deploy-mediawiki.py:244]: We should probably split this off into a generate method. Should I do a pull request for that?

File: modules/mediawiki/manifests/deploy.pp [hunk 8]
```diff
@@ -40,6 +40,16 @@
         }
     }
 
+    ensure_packages(
+        'langcodes',
```
Comment (RhinosF1) [inline modules/mediawiki/manifests/deploy.pp:44]: Is there any way to pin this?

File: modules/mediawiki/files/bin/deploy-mediawiki.py [hunk 9]
```diff
@@ -149,11 +150,15 @@ def _construct_rsync_command(time: str, dest: str, recursive: bool = True, local
...
+def _construct_git_pull(repo: str, submodules: bool = False, branch: Optional[str] = None) -> str:
     if submodules:
         extrap = '--recurse-submodules'
     else:
         extrap = ''
+
+    if branch:
+        extrap += f' --branch={branch}'
+
     return f'sudo -u {DEPLOYUSER} git -C {_get_staging_path(repo)} pull {extrap} --quiet'
```
Comment (RhinosF1) [inline modules/mediawiki/files/bin/deploy-mediawiki.py:162]: Note to self: make this method more extendable. It's going to be rubbish if we need to account for double spaces for every command in tests.

**Review by RhinosF1 — CHANGES_REQUESTED** [review]
1 question, 1 to do for any point & 1 I assume you'll want me to do.

**Review by Reception123 — APPROVED** [review]

**Comment by RhinosF1** [comment]
I filed https://phabricator.miraheze.org/T9310 as a follow up

**Pull request merged**